src/util/metadata/side_metadata/global.rs

```diff
@@ -97,6 +99,515 @@ impl SideMetadataSpec {
+        // with the actual side metadata.
+        // To achieve this, we need to apply a lock when we access side metadata. This will hide some concurrency bugs,
+        // but makes it possible for us to assert our side metadata implementation is correct.
+        #[cfg(feature = "extreme_assertions")]
+        let _lock = sanity::SANITY_LOCK.lock().unwrap();
+
+        // A few checks
+        #[cfg(debug_assertions)]
+        {
+            self.assert_value_type::<T>();
+            self.assert_metadata_mapped(data_addr);
+        }
+
+        // Actual access to the side metadata
+        let ret = access_func();
+
+        // Verifying the side metadata: checks the result with the sanity table, or store some results to the sanity table
+        verify_func(ret);
+
+        ret
+    }
+
+    /// Non-atomic load of metadata.
+    ///
+    /// # Safety
+    ///
+    /// This is unsafe because:
+    ///
+    /// 1. Concurrent access to this operation is undefined behaviour.
+    /// 2. Interleaving Non-atomic and atomic operations is undefined behaviour.
+    #[inline(always)]
+    pub unsafe fn load<T: MetadataValue>(&self, data_addr: Address) -> T {
+        self.side_metadata_access::<T, _, _, _>(
+            data_addr,
+            || {
+                let meta_addr = address_to_meta_address(self, data_addr);
+                let bits_num_log = self.log_num_of_bits;
+                if bits_num_log < 3 {
+                    let lshift = meta_byte_lshift(self, data_addr);
+                    let mask = meta_byte_mask(self) << lshift;
+                    let byte_val = meta_addr.load::<u8>();
+
+                    FromPrimitive::from_u8((byte_val & mask) >> lshift).unwrap()
+                } else {
+                    meta_addr.load::<T>()
+                }
+            },
+            |_v| {
+                #[cfg(feature = "extreme_assertions")]
+                sanity::verify_load(self, data_addr, _v);
+            },
+        )
+    }
+
+    /// Non-atomic store of metadata.
+    ///
+    /// # Safety
+    ///
+    /// This is unsafe because:
+    ///
+    /// 1. Concurrent access to this operation is undefined behaviour.
+    /// 2. Interleaving Non-atomic and atomic operations is undefined behaviour.
+    #[inline(always)]
+    pub unsafe fn store<T: MetadataValue>(&self, data_addr: Address, metadata: T) {
+        self.side_metadata_access::<T, _, _, _>(
+            data_addr,
+            || {
+                let meta_addr = address_to_meta_address(self, data_addr);
+                let bits_num_log = self.log_num_of_bits;
+                if bits_num_log < 3 {
+                    let lshift = meta_byte_lshift(self, data_addr);
+                    let mask = meta_byte_mask(self) << lshift;
+                    let old_val = meta_addr.load::<u8>();
+                    let new_val = (old_val & !mask) | (metadata.to_u8().unwrap() << lshift);
```

Comment:
`metadata` is not masked here.  It is an error if the caller passes a `metadata` value that has more than `bits_num_log` bits.  I recommend adding a `debug_assert!` to ensure `metadata` does not have more bits than necessary.